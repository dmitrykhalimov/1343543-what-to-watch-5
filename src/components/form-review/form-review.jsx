import React, {useState} from "react";
import PropTypes from "prop-types";
import {addComment} from "../../store/api-actions";
import {connect} from "react-redux";
import ErrorPopup from "../error-popup/error-popup";
import {extend} from "../../utils/utils";

const RATING_QUANTITY = 5;

const FormReview = (props) => {

  const {onReviewSubmit, id} = props;

  const initialState = {
    errorMessage: null,
    rating: null,
    comment: `1`,
    isBlocked: false,
  };

  const [currentState, setState] = useState(initialState);

  const handleRatingChange = (evt) => {
    setState(extend(currentState, {rating: Number(evt.target.value)}));
  };

  const handleCommentChange = (evt) => {
    setState(extend(currentState, {comment: evt.target.value}));
  };

  const handleFormSubmit = (evt) => {
    evt.preventDefault();
    setState(extend(currentState, {isBlocked: true}));
    onReviewSubmit(
        Number(id),
        {
          rating: currentState.rating,
          comment: currentState.comment
        },
        handleError
    );
  };

  const handleError = (message) => {
    setState(extend(currentState, {errorMessage: message}));
  };

  const handleErrorClose = () => {
    setState(extend(currentState, {errorMessage: null}));
  };

  return (
    <div className="add-review">
      {currentState.errorMessage
        ? <ErrorPopup
          errorMessage = {currentState.errorMessage}
          onCloseButtonClick = {handleErrorClose}/>
        : ``}
      <form action="#" className="add-review__form" onSubmit={handleFormSubmit}>
        <fieldset style={{border: `none`}} disabled={currentState.isBlocked}>
          <div className="rating">
            <div className="rating__stars">
              {Array(RATING_QUANTITY)
                .fill(``)
                .map((item, index) => {
                  const mark = index + 1;
                  return (
                    <React.Fragment key={index}>
                      <input
                        className="rating__input"
                        id={`star-${mark}`}
                        type="radio"
                        name="rating"
                        value={mark}
                        checked={currentState.rating === mark}
                        onChange={handleRatingChange}
                      />
                      <label className="rating__label" htmlFor={`star-${mark}`}>{`star-${mark}`}</label>
                    </React.Fragment>
                  );
                })
              }
            </div>
          </div>

          <div className="add-review__text">
            <textarea className="add-review__textarea" name="review-text" id="review-text" placeholder="Review text" onChange={handleCommentChange} value={currentState.comment}></textarea>
            <div className="add-review__submit">
              <button
                className="add-review__btn"
                type="submit"
                disabled={!(currentState.comment.length >= 50 && currentState.comment.length <= 400 && currentState.rating !== null)}
              >Post</button>
            </div>
          </div>
          {currentState.rating === null
            ? <>
            <p>Необходимо выставить оценку </p>
            </>
            : ``}

          {!(currentState.comment.length >= 50 && currentState.comment.length <= 400)
            ? <>
            <span>Отзыв должен содержать не менее 50 и не более 400 симоволов. </span>
            </>
            : ``}

          <span>Символов в отзыве: {currentState.comment.length}</span>
        </fieldset>
      </form>
    </div>
  );
};

FormReview.propTypes = {
  onReviewSubmit: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  onReviewSubmit(id, userData, handleError) {
    dispatch(addComment(id, userData, handleError));
  },
});

export {FormReview};
export default connect(null, mapDispatchToProps)(FormReview);
