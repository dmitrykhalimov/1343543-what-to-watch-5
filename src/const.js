export const ratingRanks = [
  {
    'Bad': 2.9
  },
  {
    'Normal': 4.9
  },
  {
    'Good': 7.9
  },
  {
    'Very good': 9.9
  },
  {
    'Awesome': 10
  }
];
