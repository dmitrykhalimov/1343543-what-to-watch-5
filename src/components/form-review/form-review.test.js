import React from "react";
import renderer from "react-test-renderer";
import {FormReview} from "./form-review";
import {TEST_MOCKS} from "../../const";

describe(`FormReview`, () => {
  it(`Should FormReview render correctly`, () => {
    const tree = renderer
    .create(
        <FormReview
          onReviewSubmit={TEST_MOCKS.noop}
          id={String(TEST_MOCKS.id)}
        />
    )
  .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
