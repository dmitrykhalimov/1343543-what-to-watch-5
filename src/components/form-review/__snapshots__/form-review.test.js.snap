// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormReview Should FormReview render correctly 1`] = `
<div
  className="add-review"
>
  
  <form
    action="#"
    className="add-review__form"
    onSubmit={[Function]}
  >
    <div
      className="rating"
    >
      <div
        className="rating__stars"
      >
        <input
          checked={false}
          className="rating__input"
          id="star-1"
          name="rating"
          onChange={[Function]}
          type="radio"
          value={1}
        />
        <label
          className="rating__label"
          htmlFor="star-1"
        >
          star-1
        </label>
        <input
          checked={false}
          className="rating__input"
          id="star-2"
          name="rating"
          onChange={[Function]}
          type="radio"
          value={2}
        />
        <label
          className="rating__label"
          htmlFor="star-2"
        >
          star-2
        </label>
        <input
          checked={true}
          className="rating__input"
          id="star-3"
          name="rating"
          onChange={[Function]}
          type="radio"
          value={3}
        />
        <label
          className="rating__label"
          htmlFor="star-3"
        >
          star-3
        </label>
        <input
          checked={false}
          className="rating__input"
          id="star-4"
          name="rating"
          onChange={[Function]}
          type="radio"
          value={4}
        />
        <label
          className="rating__label"
          htmlFor="star-4"
        >
          star-4
        </label>
        <input
          checked={false}
          className="rating__input"
          id="star-5"
          name="rating"
          onChange={[Function]}
          type="radio"
          value={5}
        />
        <label
          className="rating__label"
          htmlFor="star-5"
        >
          star-5
        </label>
      </div>
    </div>
    <div
      className="add-review__text"
    >
      <textarea
        className="add-review__textarea"
        id="review-text"
        name="review-text"
        onChange={[Function]}
        placeholder="Review text"
        value=""
      />
      <div
        className="add-review__submit"
      >
        <button
          className="add-review__btn"
          type="submit"
        >
          Post
        </button>
      </div>
    </div>
  </form>
</div>
`;
